8209788: Left/Right/Ctrl+A keys not working in editor of ComboBox if popup showing

arapte opened the pull request:
The issue occurs because the key events are consumed by the `ListView` in `Popup`, which displays the items.
This is a regression of [JDK-8077916](https://bugs.openjdk.java.net/browse/JDK-8077916). This change aadded several `KeyMapping`s for focus traversals to `ListView`, which consume the `Left`, `Right` and `Ctrl+A` key events.

Fix:
1. Remove the four focus traversal arrow `KeyMapping`s from `ListViewBehavior`.
2. Add the `Ctrl + A` `KeyMapping` to `ListViewBehavior` only if the `ListView`'s selection mode is set to `SelectionMode.MULTIPLE`.  `ComboBox` uses the `ListView` with `SelectionMode.SINGLE` mode.

Change unrelated to fix:
`ComboBoxListViewBehavior` adds `KeyMapping` for `Up` and `Down` keys, which are not invoked when the `ComboBox` popup is showing. When the popup is shown, the Up and Down key events are handled by the `ListView` and the `KeyMapping` code from `ComboBoxListViewBehavior` does not get executed. These KeyMapping are removed.
However this change is not needed for the fix. But this seems to be dead code. 

Verification:
Added new unit tests to verify the change. 
Also verified that the behavior ListView behaves same before and after this change.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8209788](https://bugs.openjdk.java.net/browse/JDK-8209788): Left/Right/Ctrl+A keys not working in editor of ComboBox if popup showing


### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Jeanette Winzenburg](https://openjdk.java.net/census#fastegal) (@kleopatra - Committer)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/172/head:pull/172`
`$ git checkout pull/172`

kevinrushforth commented:
/reviewers 2

abhinayagarwal commented:
Hi @arapte,

Great work on the PR :+1: 

Don't you think that all the changes in `ListViewSkin` can be moved to `ListViewBehavior`? All that we do in the skin class is to call `ListViewBehavior#updateSelectionModeKeyMapping`, which smells like feature envy. 

Moreover, `ListViewBehavior` already has change listener attached to `selectionModelProperty`, waiting for us to re-use it :wink:

kleopatra commented:
> Don't you think that all the changes in `ListViewSkin` can be moved to `ListViewBehavior`? All that we do in the skin class is to call `ListViewBehavior#updateSelectionModeKeyMapping`, which smells like feature envy.
> 
> Moreover, `ListViewBehavior` already has change listener attached to `selectionModelProperty`, waiting for us to re-use it 😉

good point :) Though - I don't like listeners to control properties in behaviors, and much less listeners to path properties (they tend to not getting cleaned on dispose). 

In the particular case of behaviors of controls with selectionModels they do (must?) because the selectionModel is not api complete (having no notion of anchor), so they jump in to cover up. Hopefully that design flaw will be fixed at some time in future, which would remove the existing listener, anyway. With just another responsibility - difference based on selectionMode - such cleanup would be harder.

Here the basic approach is to add/remove a keyMapping for multiple/single selection. Compared to current state, there's a subtle side-effect (the event bubbles up if the mapping if removed). We can achieve the same behavior (with the same side-effect) by making the mapping consume depending on whether it is handled (to select all) or not.

In code that would be pattern like:

    // in constructor

        KeyMapping selectAllMapping;
        addDefaultMapping(listViewInputMap,
              ...
             selectAll = new KeyMapping(new KeyBinding(A).shortcut(), this:: selectAll),
             ...
         };
         selectAllMapping.setAutoConsume(false);

    // selectAll with modified signature
    /**
      * Calls selectAll on selectionModel and consumes the event, if 
      * the model is available and in multimode, 
      * does nothing otherwise. 
      */
    private void selectAll(KeyEvent key) {
        MultipleSelectionModel<T> sm = getNode().getSelectionModel();
       // do nothing, let it bubble up
        if (sm == null || sm.getSelectionMode() == SelectionMode.SINGLE) return;
       // handle and consume
        sm.selectAll();
        key.consume();
    }

BTW, there are other keys that don't work as expected (from the perspective of the editor in the combo): f.i. shift-home/end is mapped to scrollToFirst/LastRow - that's hampering ux if f.i. the user has typed some input, uses them and sees her input lost because first/last row is selected. Sry to  not have noticed earlier in my bug report :(

So whatever approach we choose (mappings being removed/added or their handlers not/consuming doesn't matter), we would have to do it for several keys. Plus we have the side-effect mentioned above. The mass of change _for all listviews_ has a certain risk of breaking existing code. Think f.i. global accelerators that might (or not) get triggered depending on selection mode.

On the other hand, different mappings are needed only when the list resides in the combo's popup (and probably only if the combo is editable, didn't dig though). An alternative might be a different  inputMap (or containing different mappings) when used in combo's popup (which is similar to what Swing/X does .. no wonder I would prefer it :)

arapte commented:
> `ListViewBehavior` already has change listener attached to `selectionModelProperty`, waiting for us to re-use it

Hi @abhinayagarwal, Thanks for the suggestion. This sound good to me too. I shall make this change in next commit.

abhinayagarwal commented:
> Compared to current state, there's a subtle side-effect (the event bubbles up if the mapping if removed). We can achieve the same behavior (with the same side-effect) by making the mapping consume depending on whether it is handled (to select all) or not.

Changing the behavior inside `selectAll` is also a good option (and much cleaner as well)

> So whatever approach we choose (mappings being removed/added or their handlers not/consuming doesn't matter), we would have to do it for several keys. Plus we have the side-effect mentioned above. The mass of change for all listviews has a certain risk of breaking existing code. Think f.i. global accelerators that might (or not) get triggered depending on selection mode.

It's sad that we can't easily override (InputMapping created in) `ListViewBehavior` :(

arapte commented:
> An alternative might be a different inputMap (or containing different mappings) when used in combo's popup (which is similar to what Swing/X does .. no wonder I would prefer it :)

Thanks for the suggestion 👍 , I shall try this approach and update the PR. I am not sure if we already do this for any other control. Do you know any, if we do ?  Not actively working on this issue, Will soon get back on this :)

kleopatra commented:
the nearest to different input maps based on control state might be in listViewBehavior itself: it has differrent child maps for vertical/horizontal orientation. Could be possible to widen that a bit with another child map for vertical and in combo popup (provided it has a means to decide being in such a state for the sake of an interceptor, without api change that might be a simple entry in its properties)

arapte commented:
Please review the updated change:

1. Changed the approach to add a property named `removeKeyMappingsForComboBoxEditor` to ListView when creating it from ComboBox. Please do suggest if this name sounds Ok.
2. When this property is present some KeyMappings that are needed for ComboBox Editor get removed from `ListViewBehavior`.
3. I have considered to remove some more KeyMappings which seem like they should be passed on to ComboBox's Editor.(method `removeKeyMappingsForComboBoxEditor()`)
4. There is an existing issue with `BehaviorBase`: [JDK-8250807](https://bugs.openjdk.java.net/browse/JDK-8250807).
Due to this issue, KeyMappings from child `InputMap` do not get removed. So the keys CTRL+SHIFT+HOME and CTRL+SHIFT+END will still not work with ComboBox editor.
5. Updated ComboBoxTest for additional keys.
6. Keeping the tests added for ListView, as they seem reasonable and not present already.

Change is very specific to ComboBox editor, so it should not affect any other tests.

kleopatra commented on modules/javafx.controls/src/main/java/javafx/scene/control/skin/ListViewSkin.java:
```diff
@@ -205,6 +205,10 @@ public ListViewSkin(final ListView<T> control) {
         behavior.setOnSelectNextRow(() -> onSelectNextCell());
         behavior.setOnScrollPageDown(this::onScrollPageDown);
         behavior.setOnScrollPageUp(this::onScrollPageUp);
+        Boolean isRemoveKeyMappings = (Boolean) control.getProperties().get("removeKeyMappingsForComboBoxEditor");
+        if (isRemoveKeyMappings != null && isRemoveKeyMappings) {
+            behavior.removeKeyMappingsForComboBoxEditor();
+        }
```
hmm .. I would expect the ListViewBehavior to install the per-context mappings (vs. the ListViewSkin removing those that don't fit)

- comboBoxListViewSkin sets the context property to indicate that the list is part of the drop-down
- ListViewBehavior installs (or not) the mappings that it needs in the context as appropriate (which would remove the need for the removeKeyMappings method to be public)
- ListViewSkin has nothing special to do

What do I overlook?

arapte replied:
>     * ComboBoxListViewSkin sets the context property to indicate that the list is part of the drop-down

Property is named such that it does not show relationship but only indicates the action that is needed. `ComboBox` already adds one such property named `selectFirstRowByDefault`. If we add a property to indicate that the ListView is part of ComboBox then we might need to remove the existing property as well, and combine the actions together under one property. So seems like naming the property to indicate actions is consistent.



> ListViewBehavior to install the per-context mappings (vs. the ListViewSkin removing those that don't fit)

Installing(Not installing) specific KeyMappings has one difference compared to removing few later that, changes for installing specific mappings would be spread: 1) The KeyMappings in `FocusTraversalInputMap` will have to be broken into two parts like in the first commit. 2) In `ListViewBehavior` the default(line 82) and vertical orientation(line 130) KeyMappings will have to be broken into separate calls. Since this change is specifically for ComboBox , the regular code can be left as is and this fix can be treated as special case and kept separate packed into single call.



>     * remove the need for the removeKeyMappingsForComboBoxEditor method to be public

Keeping `ListViewSkin` untouched and making `removeKeyMappingsForComboBoxEditor()` private seems good to me too. I clearly overlooked. I shall move the change from `ListViewSkin` to end of `ListViewBehavior` constructor.

kleopatra replied:
good that we agree on where to put it :) 

Will have to think about about the remove installed mappings vs. change existing code to not installing. Right now, feels a bit smelly .. but I see the constraints (and might agree :) Will come back!

kleopatra replied:
btw, using a key-value pair in properties is just fine with me (might have been unclear, sry) - nothing else to do short of changing api (which feels too heavy weight)

One nit-pit for checking for its existence, an alternative slightly simplified pattern might be

    if (Boolean.TRUE.equals(control.getProperties().get(....)) {
        removeMappings(...);
    }

kleopatra replied:
okay, I still favor the not-adding-if-not-needed approach - and it's not _that_ much of a change to existing code. Basically, in the constructor of ListViewBehavior

- in a first block, do not add the mappings that (currently) are removed again if the  combo flag is set
-in a second block,  add the mappings for a stand-alone listView to all inputMaps as needed

code snippets:

        // not needed if in combo popup
        //  addDefaultMapping(listViewInputMap, FocusTraversalInputMap.getFocusTraversalMappings());
        addDefaultMapping(listViewInputMap,
                // not needed if in combo popup
                // new KeyMapping(HOME, e -> selectFirstRow()),
                ...
               // add those that are always needed
        }

        // same for child input maps
        addDefaultMappings(verticalInputMap, ...

        // add those for stand-alone listView
        // handle mappings that are needed only in a stand-alone listView
        if (!Boolean.TRUE.equals(control.getProperties().get("removeKeyMappingsForComboBoxEditor"))) {
            addDefaultMapping(listViewInputMap, FocusTraversalInputMap.getFocusTraversalMappings());
            addDefaultMapping(listViewInputMap, 
                    new KeyMapping(HOME, e -> selectFirstRow()),
                    new KeyMapping(END, e -> selectLastRow()),
                    new KeyMapping(new KeyBinding(HOME).shift(), e -> selectAllToFirstRow()),
                    new KeyMapping(new KeyBinding(END).shift(), e -> selectAllToLastRow()),
                    new KeyMapping(new KeyBinding(A).shortcut(), e -> selectAll()),
                    new KeyMapping(new KeyBinding(HOME).shortcut(), e -> focusFirstRow()),
                    new KeyMapping(new KeyBinding(END).shortcut(), e -> focusLastRow())
            );

            addDefaultMapping(verticalListInputMap, 
                    new KeyMapping(new KeyBinding(HOME).shortcut().shift(), e -> discontinuousSelectAllToFirstRow()),
                    new KeyMapping(new KeyBinding(END).shortcut().shift(), e -> discontinuousSelectAllToLastRow())
            );
            
        }

Tests are passing as expected, and the [open issue JDK-8250807](https://bugs.openjdk.java.net/browse/JDK-8250807) doesn't stand in the way :)

As to testing, independent of which approach is chosen at the end: I would suggest to not only test the macroscopic behavior (look if key events on combo/listView have the expected effect) but also test the base layer, that is whether or not the mappings are not/added to the inputMaps. 
 
One open (maybe not-an) issue might be the handling of the horizontal map which currently is not touched (but probably will, once the deep removal of [JDK-8250807](https://bugs.openjdk.java.net/browse/JDK-8250807) is fixed) : while not very probable, client code might change the listView in the popup to be horizontal. What would be the expected UI in that case? Anyway, could be left open for the scope of this issue, IMO.

arapte replied:
> okay, I still favor the not-adding-if-not-needed approach

I have updated the PR to use this approach. I did not think that all of `FocusTraversalInputMap.getFocusTraversalMappings()`  can be removed. But looks like they can be removed without any side effect. This approach looks good.
Additionally, changed name of the property as `excludeKeyMappingsForComboBoxEditor` to reflect the action.


> whether or not the mappings are not/added to the inputMaps.

Added this test for both ComboBox and ListView.



> Anyway, could be left open for the scope of this issue, IMO.

I also think the same that it can be kept out of scope for now.

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewTest.java:
```diff
@@ -1977,6 +2049,47 @@ public void testEventIndicesOnSelectRange() {
         assertEquals("List item at index 2 should be selected", 2, (int) sm.getSelectedIndices().get(1));
     }
 
+    @Test public void testExcludeKeyMappingsForComboBoxEditor() {
+        ListView<String> listView = new ListView<>(FXCollections
+                .observableArrayList("Item1", "Item2"));
+        listView.setCellFactory(TextFieldListCell.forListView());
```
hmm .. why the textFieldListCell? It's just a plain listCell if not in editing state .. or not?

arapte replied:
It is effect of copy paste ;), Test behaves as expected even without it, so removed.

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ComboBoxTest.java:
```diff
@@ -1334,6 +1341,160 @@ public void defaultConverterCanHandleIncorrectType_2() {
         sl.dispose();
     }
 
+    @Test public void test_EditorKeyInputsWhenPopupIsShowing() {
```
minor nit: naming is inconsistent to the other added test below (testExcludeKeyMappingsForComboBoxEditor) - either use an underscore or not (my personal preference is to not use them at all, following general java naming conventions also in tests .. but definitely not overly important :)

arapte replied:
Removed _, changed to `testEditorKeyInputsWhenPopupIsShowing`

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewTest.java:
```diff
@@ -1465,6 +1506,37 @@ private void test_rt_39559(boolean useSMSelectAll) {
         sl.dispose();
     }
 
+    @Test public void test_switchingSelectionMode() {
```
if you prefer not to, the capitalizing should be consistent - here it's test_switch, above it's test_Switch :)

arapte replied:
Changed name to `testCtrlAWhenSwitchingSelectionMode`, It sounds much better.

kleopatra reviewed (changes requested):
fix and tests look okay (added minor inline comments), verified that the tests for the fix are failing before and passing after, those added for completeness are fine also. 

As noted in one of my comments (again? :), I don't like underscores .. not even in test methods - but as they are wide spread nothing to really complain about: just be consistent with yourself :)

arapte commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewTest.java:
```diff
@@ -1465,6 +1506,37 @@ private void test_rt_39559(boolean useSMSelectAll) {
+        assertEquals(1, sm.getSelectedItems().size());
+        keyboard.doKeyPress(KeyCode.A, KeyModifier.getShortcutKey());
+        assertEquals(4, sm.getSelectedItems().size());
+
```
I had kept it with initial version of patch to test the listeners, But the patch has evolved to be very different now. But still have removed only one block. Now there are three blocks and `sm.getSelectedItems().size()` varies in each block.

kleopatra replied:
ahh .. that's what I overlooked - thanks for opening my eyes :)

arapte commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewTest.java:
```diff
@@ -171,6 +178,40 @@
         assertSame(sm, sm);
     }
 
+    @Test public void test_SwitchingSelectionModel() {
```
Done, removed _ from all the test names.

kleopatra reviewed (approved):
looks good :)

kevinrushforth commented:
The fix looks good for an editable ComboBox.

If the ComboBox is not editable, it will have the effect of making the HOME and END keys no-ops, which is a (possibly unwanted) change in behavior. I checked a couple native Windows apps and they have the behavior I would expect: the arrow keys, and the HOME / END keys navigate the text field for editable combo boxes. HOME and END go to the beginning or end of the list for non-editable combo boxes.

While we could treat that as a follow-up issue, it would be worth thinking about whether we could limit the change to editable combo boxes.

kleopatra commented:
> 
> 
> If the ComboBox is not editable, it will have the effect of making the HOME and END keys no-ops, which is a (possibly unwanted) change in behavior. I checked a couple native Windows apps and they have the behavior I would expect: the arrow keys, and the HOME / END keys navigate the text field for editable combo boxes. HOME and END go to the beginning or end of the list for non-editable combo boxes.
> 
> While we could treat that as a follow-up issue, it would be worth thinking about whether we could limit the change to editable combo boxes.

outch .. how did I overlook that .. (seems reviewing doesn't belong to my strengths ;)

This fix should not break (correct) existing behavior, so back to thinking ..

arapte commented:
> If the ComboBox is not editable, it will have the effect of making the HOME and END keys no-ops, which is a (possibly unwanted) change in behavior.

I have updated PR with changes for non editable ComboBox.
I could think of adding another property to propagate the editable property of ComboBox to ListViewBehavior. So now this fix adds another property `editableComboBoxEditor`, not sure if there is other way to handle it.
The change adds HOME and END KeyMappings when ComboBox is non editable and removes them when ComboBox is editable.
If the change sounds Ok, I shall include test in next commit.

Also, there is one change in the if condition that was suggested by Jeanette before,
`if (!Boolean.TRUE.equals(control.getProperties().containsKey("excludeKeyMappingsForComboBoxEditor")))`
is changed to,
`if (Boolean.FALSE.equals(control.getProperties().containsKey("excludeKeyMappingsForComboBoxEditor")))`

It seems safe as `control.getProperties().containsKey()` returns either true or false.

kevinrushforth commented on modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java:
```diff
@@ -138,7 +140,8 @@ public ListViewBehavior(ListView<T> control) {
         );
 
-        if (!Boolean.TRUE.equals(control.getProperties().containsKey("excludeKeyMappingsForComboBoxEditor"))) {
+        if (Boolean.FALSE.equals(control.getProperties().containsKey("excludeKeyMappingsForComboBoxEditor"))) {
```
Unless I'm missing something, you don't need to compare with a `Boolean` at all since containsKey returns a `boolean` primitive type, so I think this can be simplified to:

```
if (!control.getProperties().containsKey("excludeKeyMappingsForComboBoxEditor")) {
```

If instead you do want to check the value of the property, and not just its existence, you would need the following, and it would be important to check `!TRUE.equals` rather than `FALSE.equals` so that an unset property would be treated as `false`.

```
if (!Boolean.TRUE.equals(control.getProperties().get("excludeKeyMappingsForComboBoxEditor"))) {
```

arapte replied:
Changed to use
`if (!control.getProperties().containsKey("excludeKeyMappingsForComboBoxEditor")) {`

kevinrushforth commented on modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java:
```diff
@@ -340,6 +354,26 @@ public void changed(
 
+    private final InvalidationListener propertiesListener = inv -> {
+        Boolean isComboBoxEditable = (Boolean)getNode().getProperties().get("editableComboBoxEditor");
+        if (isComboBoxEditable != null) {
```
Maybe simplify this as follows, to match the earlier logic?

```
if (Boolean.FALSE.equals(getNode().getProperties().get("editableComboBoxEditor"))) {
```

arapte replied:
With this change, we need to check both FALSE and TRUE.
For a Non ComboBox ListView `getNode().getProperties().get("editableComboBoxEditor")` would be `null`.

```
if (Boolean.FALSE.equals(getNode().getProperties().get("editableComboBoxEditor"))) { add KeyMappings }
else if (Boolean.TRUE.equals(getNode().getProperties().get("editableComboBoxEditor"))) { remove KeyMappings }
else { ListView is not contained by ComboBox }
```

arapte commented:
> Would it be possible to add some tests to verify the behavior of HOME and END for editable and non-editable ComboBox controls?

@kevinrushforth @kleopatra 
Please check the updated changes. _ComboBoxTest_ and _ListViewTest_ both have minor modifications in how they access _KeyMappings_. and added test for verifying **HOME** and **END** key with both editable and non editable ComboBox.

kleopatra commented:
hmm .. this is getting unwieldy, isn't it ;)

The pain points:
- cascade of listeners (editable -> comboSkin -> properties -> behavior)
- dynamic change (add/remove) of mappings
- multiple key/value pairs for basically the same - though variant - state 

My suggestion would be to take a step back (in solution path): near the beginning was the evaluation of using different inputMaps for different state contexts. Which was not further evaluated because it looked like we could get away with simply configuring the mappings - based on certain condition - once at instantiation time. Which has the advantage of not touching too much code but unfortunely turned out to be not enough. 

Meanwhile, I'm convinced that in the long run there is no way around different inputMaps based on context: the differences in behavior (stand-alone vs. editable combo-popup vs. not-editable combo-popup) are many - f.i. focus-only navigation doesn't make sense in the popup (should be selection navigation always), left/right in a not-editable should trigger selection navigation .. and certainly more. So we not only have to enable/disable certain mappings, but also re-map the triggered behavior. 

That's too broad for this issue, but we could take a step into that direction: use the InputMap/Mapping API to help - it was designed for exactly such a differentiation :) The step would be to use interceptors (instead of dynamic modification of the mappings list), they are available on both inputMap and mapping level. As a first step, we could use the latter: keep the addition of mappings as-is (before the fix) and add interceptors to mappings for inclusion/exclusion based on context. No listeners, no dynamic modification, just one marker in the properties .. hopefully :)

Raw code snippets:

    // let combo skin put a Supplier for editable as value
    getProperties().put("comboContext", (Supplier<Boolean>) () -> getSkinnable().isEditable());

    // let listView behavior use the supplier to build interceptors
    Supplier<Boolean> comboEditable = (Supplier<Boolean>) control.getProperties().get("comboContext");
    Predicate<KeyEvent> interceptIfInCombo = e -> comboEditable != null;
    Predicate<KeyEvent> interceptIfInEditableCombo = e -> comboEditable != null && comboEditable.get();
    
    if (comboEditable == null) {
        // add focus traversal mappings if not in combo popup
        addDefaultMapping(listViewInputMap, FocusTraversalInputMap.getFocusTraversalMappings());
    }
    // add mappings with appropriate interceptors
    addDefaultMapping(listViewInputMap,
        // missing api in KeyMapping: no constructor taking KeyCode and interceptor   
        new KeyMapping(new KeyBinding(HOME), e -> selectFirstRow(), interceptIfInEditableCombo),
        new KeyMapping(new KeyBinding(END), e -> selectLastRow(), interceptIfInEditableCombo),
        new KeyMapping(new KeyBinding(HOME).shift(), e -> selectAllToFirstRow(), interceptIfInCombo),
        new KeyMapping(new KeyBinding(END).shift(), e -> selectAllToLastRow(), interceptIfInCombo),
        ... 

With this, the tests for key navigation are passing, the low-level mapping tests will have to be re-formulated to test for not/intercepted vs. existence.

What do you think?

arapte commented:
@kleopatra  @kevinrushforth  
Change looks pretty with interceptor.
Please take a look.

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ComboBoxTest.java:
```diff
@@ -1334,6 +1341,177 @@ public void defaultConverterCanHandleIncorrectType_2() {
+        assertTrue(testInterceptor(verticalInputMappings, new KeyBinding(KeyCode.END).shortcut().shift()));
+    }
+
+    private boolean testInterceptor(ObservableList<?> mappings, KeyBinding binding) {
+        int i = mappings.indexOf(new KeyMapping(binding, null));
+        return ((KeyMapping)mappings.get(i)).getInterceptor().test(null);
+    }
+
```
his throws an NPE without the fix - which makes the test using this method still fail (good!), but a bit unspecific for my taste. A slight re-organisation of the helpers might help: move the asserts down instead of returning a boolean.

arapte replied:
Change function as you suggested. Unlike ListViewTest, ComboBoxTest needs to verify if the interceptor correctly evaluates to either true or false, so it needed a minor tweak to use `assertEquals()`

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewTest.java:
```diff
@@ -1977,6 +2045,44 @@ public void testEventIndicesOnSelectRange() {
+        sl.dispose();
+    }
+
+    private boolean testInterceptor(ObservableList<?> mappings, KeyBinding binding) {
+        int i = mappings.indexOf(new KeyMapping(binding, null));
+        return ((KeyMapping)mappings.get(i)).getInterceptor().test(null);
```
same as for ComboBoxTest

kleopatra reviewed (changes requested):
Fix looks fine and indeed pretty :) Verified tests failing before and after the fix.

Left some minor comments inline.

arapte commented on modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java:
```diff
@@ -78,12 +80,19 @@ public ListViewBehavior(ListView<T> control) {
+                (Supplier<Boolean>) control.getProperties().get("editableComboBoxEditor");
+        Predicate<KeyEvent> pIsInComboBox = e -> isListViewOfComboBox != null;
+        Predicate<KeyEvent> pIsInEditableComboBox =
+                e -> isListViewOfComboBox != null && isListViewOfComboBox.get();
```
Even I was bit skeptical about prefix p, before it those names sounded like a boolean. So I wanted to give it a different name. But as you said, isIn/Editable/Combo looks enough expressive. Changed names.

arapte commented on modules/javafx.controls/src/main/java/javafx/scene/control/skin/ComboBoxListViewSkin.java:
```diff
@@ -504,6 +505,7 @@ private void updateCellFactory() {
             {
                 getProperties().put("selectFirstRowByDefault", false);
+                getProperties().put("editableComboBoxEditor", (Supplier<Boolean>) () -> getSkinnable().isEditable());
             }
```
Changed to `editableComboBox` and added a short two liner comment about it.

arapte commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ComboBoxTest.java:
```diff
@@ -1334,6 +1341,177 @@ public void defaultConverterCanHandleIncorrectType_2() {
         sl.dispose();
     }
 
+    @Test public void testEditorKeyInputsWhenPopupIsShowing() {
+        final ComboBox<String> cb = new ComboBox<>(FXCollections.observableArrayList("a", "b", "c"));
+        cb.setEditable(true);
+        StageLoader sl = new StageLoader(cb);
```
I did not understand what low-level test mean here. But have added test for navigation keys which verify using `SelectionModel.getSelectedIndex()`. Please point if it is sufficient.

kleopatra replied:
my apologies for having been unclear: low-level == implementation details in keyMappings whereas functional are those testing the effect on firing keys. 

But you guessed so much and added tests of key-firing on not-editable combo, so looks good now :)

arapte commented:
> Left some minor comments inline.

Updated PR with corrections, please take a look.

kleopatra reviewed (approved):
looks good to me now :)

arapte commented:
/integrate
